diff -p -up cluster-2.03.11/cman/init.d/cman.in.orig cluster-2.03.11/cman/init.d/cman.in
--- cluster-2.03.11/cman/init.d/cman.in.orig	2009-01-22 14:33:51.000000000 +0200
+++ cluster-2.03.11/cman/init.d/cman.in	2009-04-07 15:20:09.000000000 +0200
@@ -107,7 +107,7 @@ start_ccsd()
     status ccsd &> /dev/null
     if [ $? -ne 0 ]
     then
-	errmsg=$(/sbin/ccsd $CCSD_OPTS 2>&1)
+	errmsg=$(/usr/sbin/ccsd $CCSD_OPTS 2>&1)
 	rtrn=$?
 	return $rtrn
     fi
@@ -117,15 +117,15 @@ start_ccsd()
 start_cman()
 {
     # cman
-    /sbin/cman_tool status &> /dev/null
+    /usr/sbin/cman_tool status &> /dev/null
     if [ $? -ne 0 ]
     then
-	errmsg=$( /sbin/cman_tool -t $CMAN_CLUSTER_TIMEOUT -w join \
+	errmsg=$( /usr/sbin/cman_tool -t $CMAN_CLUSTER_TIMEOUT -w join \
 	    $cman_join_opts 2>&1 ) || return 1
 
 	if [ $CMAN_QUORUM_TIMEOUT -gt 0 ]
 	then
-	    errmsg=$( /sbin/cman_tool -t $CMAN_QUORUM_TIMEOUT \
+	    errmsg=$( /usr/sbin/cman_tool -t $CMAN_QUORUM_TIMEOUT \
 		    -q wait 2>&1 ) || return 1
 	fi
     fi
@@ -165,32 +165,32 @@ start_daemons()
 {
     status groupd &> /dev/null
     if [ $? -ne 0 ]; then
-	errmsg=$( /sbin/groupd 2>&1 ) || return 1
+	errmsg=$( /usr/sbin/groupd 2>&1 ) || return 1
     fi
     status fenced &> /dev/null
     if [ $? -ne 0 ]; then
-	errmsg=$( /sbin/fenced 2>&1 ) || return 1
+	errmsg=$( /usr/sbin/fenced 2>&1 ) || return 1
     fi
     status dlm_controld &> /dev/null
     if [ $? -ne 0 ]; then
-	errmsg=$( /sbin/dlm_controld 2>&1 ) || return 1
+	errmsg=$( /usr/sbin/dlm_controld 2>&1 ) || return 1
     fi
     status gfs_controld &> /dev/null
     if [ $? -ne 0 ]; then
-	errmsg=$( /sbin/gfs_controld 2>&1 ) || return 1
+	errmsg=$( /usr/sbin/gfs_controld 2>&1 ) || return 1
     fi
     return 0
 }
 
 start_fence()
 {
-    /sbin/cman_tool status | grep Flags | grep 2node &> /dev/null
+    /usr/sbin/cman_tool status | grep Flags | grep 2node &> /dev/null
     if [ $? -ne 0 ]
     then
-        errmsg=$( /sbin/fence_tool -w -t $FENCED_START_TIMEOUT join \
+        errmsg=$( /usr/sbin/fence_tool -w -t $FENCED_START_TIMEOUT join \
                  > /dev/null 2>&1 ) || return 1
     else
-        errmsg=$( /sbin/fence_tool -w -t $FENCED_START_TIMEOUT \
+        errmsg=$( /usr/sbin/fence_tool -w -t $FENCED_START_TIMEOUT \
                  -m $FENCED_MEMBER_DELAY join \
                  > /dev/null 2>&1 ) || return 1
     fi
@@ -201,7 +201,7 @@ start_fence_xvmd()
 {
     status fence_xvmd &> /dev/null
     if [ $? -ne 0 ]; then
-	errmsg=$( /sbin/fence_xvmd $FENCE_XVMD_OPTS 2>&1 ) || return 1
+	errmsg=$( /usr/sbin/fence_xvmd $FENCE_XVMD_OPTS 2>&1 ) || return 1
     fi
     return 0
 }
@@ -462,10 +462,10 @@ stop_ccsd()
 
 stop_cman()
 {
-    /sbin/cman_tool status &> /dev/null
+    /usr/sbin/cman_tool status &> /dev/null
     if [ $? -eq 0 ]
     then
-    errmsg=$( /sbin/cman_tool -t $CMAN_SHUTDOWN_TIMEOUT \
+    errmsg=$( /usr/sbin/cman_tool -t $CMAN_SHUTDOWN_TIMEOUT \
 	    -w leave $1 2>&1 ) || return 1
     fi
     return 0 # all ok
@@ -492,7 +492,7 @@ stop_fence()
 {
     if /sbin/pidof fenced &> /dev/null
     then
-	/sbin/fence_tool -w leave > /dev/null 2>&1
+	/usr/sbin/fence_tool -w leave > /dev/null 2>&1
 	rtrn=$?
 	sleep 1 # A bit of time for fenced to exit
 	return $rtrn
